ibis/backends/dask/tests/test_udf.py

```diff
@@ -0,0 +1,483 @@
+    expr = my_string_length(t.a)
+
+    assert isinstance(expr, ir.ColumnExpr)
+
+    result = expr.execute()
+    expected = df.a.str.len().mul(2).compute()
+
+    tm.assert_series_equal(result, expected, check_names=False)
+
+
+def test_elementwise_udf_with_non_vectors(con):
+    expr = my_add(1.0, 2.0)
+    result = con.execute(expr)
+    assert result == 3.0
+
+
+def test_multiple_argument_udf(con, t, df):
+    expr = my_add(t.b, t.c)
+
+    assert isinstance(expr, ir.ColumnExpr)
+    assert isinstance(expr, ir.NumericColumn)
+    assert isinstance(expr, ir.FloatingColumn)
+
+    result = expr.execute()
+    expected = (df.b + df.c).compute()
+    tm.assert_series_equal(result, expected)
+
+
+def test_multiple_argument_udf_group_by(con, t, df):
+    expr = t.groupby(t.key).aggregate(my_add=my_add(t.b, t.c).sum())
+
+    assert isinstance(expr, ir.TableExpr)
+    assert isinstance(expr.my_add, ir.ColumnExpr)
+    assert isinstance(expr.my_add, ir.NumericColumn)
+    assert isinstance(expr.my_add, ir.FloatingColumn)
+
+    result = expr.execute()
+    expected = pd.DataFrame(
+        {'key': list('ab'), 'my_add': [sum([1.0 + 4.0, 2.0 + 5.0]), 3.0 + 6.0]}
+    )
+    tm.assert_frame_equal(result, expected)
+
+
+def test_udaf(con, t, df):
+    expr = my_string_length_sum(t.a)
+
+    assert isinstance(expr, ir.ScalarExpr)
+
+    result = expr.execute()
+    expected = t.a.execute().str.len().mul(2).sum()
+    assert result == expected
+
+
+def test_udaf_analytic(con, t, df):
+    expr = zscore(t.c)
+
+    assert isinstance(expr, ir.ColumnExpr)
+
+    result = expr.execute()
+
+    def f(s):
+        return s.sub(s.mean()).div(s.std())
+
+    expected = (f(df.c)).compute()
+    tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.xfail(
+    raises=NotImplementedError, reason='TODO - windowing - #2553'
+)
+def test_udaf_analytic_groupby(con, t, df):
+    expr = zscore(t.c).over(ibis.window(group_by=t.key))
+
+    assert isinstance(expr, ir.ColumnExpr)
+
+    result = expr.execute()
+
+    def f(s):
+        return s.sub(s.mean()).div(s.std())
+
+    expected = df.groupby('key').c.transform(f)
+    tm.assert_series_equal(result, expected)
+
+
+def test_udaf_groupby(df3):
+    con = connect({'df3': df3})
+    t = con.table('df3')
+
+    expr = t.groupby(t.key).aggregate(my_corr=my_corr(t.a, t.b))
+
+    assert isinstance(expr, ir.TableExpr)
+
+    result = expr.execute().sort_values('key').reset_index()
+
+    dfi = df3.set_index('key').compute()
+    expected = pd.DataFrame(
+        {
+            'key': list('def'),
+            'my_corr': [
+                dfi.loc[value, 'a'].corr(dfi.loc[value, 'b'])
+                for value in 'def'
+            ],
+        }
+    )
+
+    columns = ['key', 'my_corr']
+    tm.assert_frame_equal(result[columns], expected[columns])
+
+
+def test_nullable():
+    t = ibis.table([('a', 'int64')])
+    assert nullable(t.a.type()) == (type(None),)
+
+
+def test_nullable_non_nullable_field():
```

Comment:
I think this test can probably be lifted to expr? Don't need you need any Dask specific things to test this.

Reply:
https://github.com/ibis-project/ibis/pull/2648#discussion_r600751733